src/main/scala/com/databricks/spark/xml/XmlInputFormat.scala

```diff
@@ -168,46 +168,64 @@ private[xml] class XmlRecordReader extends RecordReader[LongWritable, Text] {
     }
   }
 
-  /**
-   * Read until the given data are matched with `mat`.
-   * When withinBlock is true, it saves the data came in.
-   *
-   * @param mat bytes to match
-   * @param withinBlock start offset
-   * @return whether it finds the match successfully
-   */
-  private def readUntilMatch(mat: Array[Byte], withinBlock: Boolean): Boolean = {
-    var i: Int = 0
-    while(true) {
-      val b: Int = in.read
-      if (b == -1) {
-        currentStartTag = startTag
-        return false
-      }
-      if (withinBlock) {
-        buffer.write(b)
-      }
-      if (b == mat(i)) {
-        i += 1
-        if (i >= mat.length) {
-          currentStartTag = startTag
-          return true
+  private def readUntilStartElement(): Boolean = {
+    @annotation.tailrec
+    def loop(i: Int): Boolean = {
```

Comment:
If this were my private repo, I'd be happy to accept this PR as it is but I am worried because this one is now under Databricks (and I am actually not a Databricks guy but just an individual). So, please do me a favour to stay in a safe-side.

Also, actually. I am about to release a new one. So, let's use loop one for now and then include that in the release. I would like to include this change. After that, maybe we can make another PR to use recursive one with cc'ing other guys to discuss further.